workers/unity/Packages/io.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Partials/Improbable.Gdk.TransformSynchronization.CompressedQuaternion

```diff
@@ -0,0 +1,215 @@
+    var q = QuaternionPool.BorrowAndInitialize(quaternion);
+
+    // Iterate through quaternion to find the index of the largest component.
+    uint largestIndex = 0;
+    var largestValue = Mathf.Abs(q[largestIndex]);
+    for (uint i = 1; i < 4; ++i)
+    {
+        var componentAbsolute = Mathf.Abs(q[i]);
+        if (componentAbsolute > largestValue)
+        {
+            largestIndex = i;
+            largestValue = componentAbsolute;
+        }
+    }
+
+    // Since -q == q, transform the quaternion so that the largest component is positive. This means the sign
+    // bit of the largest component does not need to be sent.
+    uint negativeBit = quaternion[(int) largestIndex] < 0 ? 1u : 0u;
+
+    // Initialise a uint with the index of the largest component. The variable is shifted left after each
+    // section of the uint is populated. At the end of the loop, the uint has the following structure:
+    // |largest index (2)|signbit (1) + component (9)|signbit (1) + component (9)|signbit (1) + component (9)|
+    uint compressedQuaternion = largestIndex;
+    for (uint i = 0; i < 4; i++)
+    {
+        if (i != largestIndex)
+        {
+            // If quaternion needs to be transformed, flip the sign bit.
+            uint signBit = (q[i] < 0 ? 1u : 0u) ^ negativeBit;
+
+            // The maximum possible value of the second largest component in a unit quaternion is 1/sqrt(2), so
+            // translate the value from the range [0, 1/sqrt(2)] to [0, 1] for higher precision. Add 0.5f for
+            // rounding up the value before casting to a uint.
+            uint magnitude = (uint) (((1u << 9) - 1u) * (Mathf.Abs(q[i]) / SqrtHalf) + 0.5f);
+
+            // Shift uint by 10 bits then place the component's sign bit and 9-bit magnitude in the gap.
+            compressedQuaternion = (compressedQuaternion << 10) | (signBit << 9) | magnitude;
+        }
+    }
+
+    // Return array to pool.
+    QuaternionPool.ReturnArray(q);
+
+    return new CompressedQuaternion(compressedQuaternion);
+}
+
+private static class QuaternionPool
+{
+    private static System.Collections.Concurrent.ConcurrentStack<float[]> pool;
+
+    public static float[] Borrow()
+    {
+        if (pool == null)
+        {
+            pool = new System.Collections.Concurrent.ConcurrentStack<float[]>();
+        }
+
+        if (pool.TryPop(out var obj))
+        {
+            return obj;
+        }
+
+        return new float[4];
+    }
+
+    public static float[] BorrowAndInitialize(Quaternion q)
+    {
+        var array = Borrow();
+        array[0] = q.x;
+        array[1] = q.y;
+        array[2] = q.z;
+        array[3] = q.w;
+        return array;
+    }
+
+    public static void ReturnArray(float[] obj)
+    {
+        pool.Push(obj);
+    }
+
+    public static Quaternion GetUnityQuaternion(float[] obj)
+    {
+        var q = new Quaternion(obj[0], obj[1], obj[2], obj[3]);
+        ReturnArray(obj);
+        return q;
+    }
+}
```

Comment:
newline

Reply:
we don't have newlines at the end of any of the other partials?

It means the generated code ends up having 2 newlines between the `}` here and the start of the `Serialization` class in the struct